### PR TITLE
*: support online backup

### LINKED ISSUE #1290: support online backup
Refer https://github.com/facebook/rocksdb/wiki/How-to-backup-RocksDB%3F

We can send SIGUSR2 signal and let RocksDB backup data.

### PR DESCRIPTION
**Pull request by overvenus**
Support online backup.

Q: How to make a backup?
A:  kill -12 `pgrep tikv-server`

Pending ngaut/rust-rocksdb#56

Close #1290 

PTAL @siddontang @BusyJay

### CODE REVIEW

**Comment by BusyJay** [comment]
I don't think it's a good idea to use signal here. Whenever online backup is required, tikv-ctl is involved normally. I propose to use tikv-ctl to send a raft command or something similar to trigger the job.

File: src/bin/tikv-server.rs [hunk 1]
```diff
@@ -584,7 +594,17 @@ fn run_raft_server(listener: TcpListener,
                           resolver,
                           snap_mgr)
         .unwrap();
-    start_server(svr, event_loop, engine);
+
+    let backup_path = {
+        let mut path = get_toml_string(config, "server.backup", Some(String::new()));
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: can we use unwrap_or_else to construct the backup path if toml has none.

**Comment by siddontang** [comment]
I think signal is ok here now. Later if we support Restful API or gRPC, we can use a tool to trigger it.

File: src/bin/tikv-server.rs (hunk 1 above)
Comment (overvenus) [inline src/bin/tikv-server.rs]: > fn get_toml_string(config: &toml::Value, name: &str, default: Option<String>) -> String 

No `unwrap_or_else` here :worried:

**Comment by BusyJay** [comment]
There are only two signals that may be customed by user, it will be very inconvenient if we want to introduce more command like consistency check, force compact or gc.

**Comment by siddontang** [comment]
Of course, the signal is not enough for user. I prefer solving these with gRPC and Restful API later.
Using signal for online backup may be the easiest way now. If we really meet another operation which needs be triggered by signal, we should consider adding command, or exposing a HTTP port for Restful API, or waiting gRPC. 

/cc @ngaut

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -584,7 +594,17 @@ fn run_raft_server(listener: TcpListener,
...
+        if path.is_empty() {
+            path = store_path;
```
Comment (zhangjinpeng1987) [inline src/bin/tikv-server.rs]: If `store_path="/somepath/data/"`, then `backup_path="/somepath/data/_backup/"`, if `store_path="/somepath/data"`, then `backup_path="/somepath/data_backup/"`, is it except?

**Comment by BusyJay** [comment]
I prefer keeping this signal to let tikv reload configuration.

File: src/bin/tikv-server.rs [hunk 3]
```diff
@@ -602,7 +612,24 @@ fn run_raft_server(listener: TcpListener,
                           resolver,
                           snap_mgr)
         .unwrap();
-    start_server(svr, event_loop, engine);
+
+    let backup_path = {
+        let mut path = get_toml_string(config, "server.backup", Some(String::new()));
+        if path.is_empty() {
+            path = store_path.to_owned();
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: the real db path is `store_path/db`, seem we can use `store_path/db_backup` directly.

File: src/bin/tikv-server.rs [hunk 4]
```diff
@@ -602,7 +612,24 @@ fn run_raft_server(listener: TcpListener,
...
+                } else {
+                    break;
+                }
+            }
+            path.push_str("_backup");
+        }
+        info!("backup path: {:?}", path);
+        path
+    };
+    start_server(svr, event_loop, engine, &backup_path);
```
Comment (hhkbp2) [inline src/bin/tikv-server.rs]: Verify `backup_path` is valid (it is valid path, a dir, and exists) before `start_server`.

File: etc/config-template.toml [hunk 5]
```diff
@@ -29,8 +29,8 @@ end-point-concurrency = 8
 # set store capacity, if no set, use unlimited or disk size later.
 # capacity = 0 # 0 is unlimited.
 
-# set backup path, if not set, use `store` with a suffix "_backup".
-backup = "/tmp/tikv/backup"
+# set backup path, if not set, use "backup" under store path.
+backup = "/tmp/tikv/store/backup"
```
Comment (siddontang) [inline etc/config-template.toml]: use empty here.

**Comment by zhangjinpeng1987** [comment]
@siddontang  Do we need restore from a backup?

**Comment by zhangjinpeng1987** [comment]
Should we support http API to do such things?

**Comment by siddontang** [comment]
Here we just want to dump the data dynamically and debug.

File: src/bin/tikv-server.rs [hunk 6]
```diff
@@ -555,6 +575,13 @@ fn handle_signal(ch: SendCh<Msg>, engine: Arc<DB>) {
                     info!("{}", v)
                 }
             }
+            SIGUSR2 => {
+                info!("making backup ...");
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: add the backup path in log

File: src/util/rocksdb.rs [hunk 7]
```diff
@@ -138,6 +138,11 @@ fn db_exist(path: &str) -> bool {
     fs::read_dir(&path).unwrap().next().is_some()
 }
 
+pub fn backup(db: &DB, path: &str) -> Result<(), String> {
```
Comment (siddontang) [inline src/util/rocksdb.rs]: add a test

**Comment by siddontang** [comment]
Rest LGTM

Now we can use SIGUSR2 for backup, if later we need using signal for other more important feature, we can remove it at that time.

**Comment by overvenus** [comment]
PTAL @BusyJay

**Comment by siddontang** [comment]
Ping @BusyJay

File: src/bin/tikv-server.rs [hunk 8]
```diff
@@ -479,22 +479,39 @@ fn build_raftkv(config: &toml::Value,
      engine)
 }
 
-fn get_store_path(matches: &Matches, config: &toml::Value) -> String {
-    let path = get_flag_string(matches, "s")
-        .unwrap_or_else(|| get_toml_string(config, "server.store", Some(TEMP_DIR.to_owned())));
-    if path == TEMP_DIR {
-        return path;
+fn get_store_and_backup_path(matches: &Matches, config: &toml::Value) -> (String, String) {
+    fn canonicalize_path(path: &str) -> String {
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: I prefer not using nested function.

File: src/bin/tikv-server.rs [hunk 9]
```diff
@@ -479,22 +479,39 @@ fn build_raftkv(config: &toml::Value,
...
+    let store_abs_path = if store_path == TEMP_DIR {
+        store_path.to_owned()
+    } else {
+        canonicalize_path(&store_path)
+    };
+
+    // Backup path
+    let mut backup_path = get_toml_string(config, "server.backup", Some(String::new()));
+    if backup_path.is_empty() {
+        backup_path = store_abs_path.clone();
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: What if `store_abs_path` is `TEMP_DIR`?

File: src/bin/tikv-server.rs [hunk 10]
```diff
@@ -479,22 +479,39 @@ fn build_raftkv(config: &toml::Value,
...
+        if !backup_path.ends_with('/') {
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Use `Path::join` instead.

File: src/bin/tikv-server.rs [hunk 11]
```diff
@@ -514,15 +534,15 @@ fn start_server<T, S>(mut server: Server<T, S>, mut el: EventLoop<Server<T, S>>,
...
         })
         .unwrap();
-    handle_signal(ch, engine);
+    handle_signal(ch, engine, backup_path);
     h.join().unwrap();
 }
 
 #[cfg(unix)]
-fn handle_signal(ch: SendCh<Msg>, engine: Arc<DB>) {
+fn handle_signal(ch: SendCh<Msg>, engine: Arc<DB>, backup_path: &str) {
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Add the extra parameter at L592 too.

File: etc/config-template.toml [hunk 12]
```diff
@@ -29,6 +29,9 @@ end-point-concurrency = 8
 # set store capacity, if no set, use unlimited or disk size later.
 # capacity = 0 # 0 is unlimited.
 
+# set backup path, if not set, use "backup" under store path.
+# backup = "/tmp/tikv/store/backup"
```
Comment (BusyJay) [inline etc/config-template.toml]: Is this the default setting?

File: src/bin/tikv-server.rs [hunk 13]
```diff
@@ -506,22 +506,41 @@ fn build_raftkv(config: &toml::Value,
...
+        .unwrap_or_else(|| get_toml_string(config, "server.store", Some(TEMP_DIR.to_owned())));
+    let store_abs_path = if store_path == TEMP_DIR {
+        TEMP_DIR.to_owned()
+    } else {
+        canonicalize_path(&store_path)
+    };
+
+    // Backup path
+    let mut backup_path = get_toml_string(config, "server.backup", Some(String::new()));
+    if backup_path.is_empty() && !store_abs_path.is_empty() {
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: You mean `store_abs_path != TEMP_DIR`?

File: etc/config-template.toml (hunk 12 above)
Comment (overvenus) [inline etc/config-template.toml]: Yes, if `store = "/tmp/tikv/store"`.

File: src/bin/tikv-server.rs (hunk 13 above)
Comment (overvenus) [inline src/bin/tikv-server.rs]: Yes, they are the same thing.

Comment (BusyJay) [inline src/bin/tikv-server.rs]: I think it's more readable to use `store_abs_path != TEMP_DIR`.

**Pull request merged**